### PR TITLE
Fix increment over time validation

### PR DESCRIPTION
**Pull request by r7vme**
Dependency for: https://github.com/giantswarm/aws-operator/pull/921

This change fixes validation of version bundles by time. Current validation that "version bundle increments over time" was not correct. Before we assumed that if semver of bundle1 bigger than semver of bundle2 (e.g. 3.1.0 > 3.0.3) than bundle time should also be bigger (newer). But we actually need take into account that patch version can be released much later than next minor/major version. For example 3.0.3 patch release of 3.0 can be released independently of 3.1 releases.

So in this change i'm taking version bundle, filter all version bundles that related to input version bundle and do "increments over time" validation. e.g 3.0.3 will be checked for 3.0.0, 3.0.1, 3.0.2, 3.0.3, but not for 2.0.1 or 3.1.0.

### CODE REVIEW

File: bundles.go [hunk 1]
```diff
@@ -103,6 +110,55 @@ func CopyBundles(bundles []Bundle) []Bundle {
 	return copy
 }
 
+// filterSameSemVerTypeBundles filters bundles that have the same semver type as
+// input bundle. SemVer type can be: Major, Minor or Patch.
+func filterSameSemVerTypeBundles(bundle Bundle, bundles []Bundle) ([]Bundle, error) {
```
Comment (r7vme) [inline bundles.go]: it's not actually same "semver type". I need to rename this function.

File: bundles.go [hunk 2]
```diff
@@ -38,12 +39,18 @@ func (b Bundles) Validate() error {
 	}
 
 	for _, v := range perProviderVersions {
-		b1 := CopyBundles(v)
```
Comment (r7vme) [inline bundles.go]: Moved to separate function `validateIncrementsOverTime` below

File: bundles.go [hunk 3]
```diff
@@ -103,6 +110,57 @@ func CopyBundles(bundles []Bundle) []Bundle {
 	return copy
 }
 
+// filterSameSeriesBundles filters bundles that are from the same semver series as
+// input bundle. For major version bundle (e.g. 3.0.0) this function will return
+// all available major bundles. For minor all minor bundles from the same major bundle.
+// For patch all patch versions from the same minor bundle.
```
Comment (r7vme) [inline bundles.go]: In this function main logic. Depending on input bundle it outputs bundles that related to the same "series" (or same release). e.g. 3.0.1 is patch release, so all 3.0 release patch releases will be returned.

File: bundles_test.go [hunk 4]
```diff
@@ -2100,7 +2219,7 @@ func Test_Bundles_Validate(t *testing.T) {
 					Deprecated:   false,
 					Name:         "kubernetes-operator",
 					Time:         time.Unix(10, 5),
-					Version:      "0.1.0",
+					Version:      "0.1.1",
```
Comment (r7vme) [inline bundles_test.go]: We now should fail only for bundles from the same series. (e.g. 0.1.1 and 0.1.0)

Comment (r7vme) [reply bundles_test.go]: we don't care if 0.0.9 was released after 0.1.0

**Comment by xh3b4sd** [comment]
If we do not care about the order in time in which bundles are created then we could actually just drop the validation in this regard and be done. I am not sure why we have additional filtering now. This looks more complicated than it has to be to me right now. Why not just dropping the time validation completely? Shouldn't this be easier?

**Comment by tuommaki** [comment]
> If we do not care about the order in time in which bundles are created then we could actually just drop the validation in this regard and be done.

If we stick to the rule that version bundles for specific `Name` (and `Provider` in some cases) can only contain one active release, then yes, I'd drop the validation as a whole, but if we provide patch updates to older version bundles (would it make sense?) then I think this is useful.

**Comment by xh3b4sd** [comment]
Patch updates to older versions make sense. We just saw that we have to branch off for reasons which is why this PR got created in the first place. I think the date validation is not useful then at all so this should be good. Note that I added a lot of validation in the first place based on plain guesses and gut feelings. So adapting when it makes sense works for me. We will also have the version design brainstorming to address certain issues this week.

**Comment by xh3b4sd** [comment]
And I think we can close this here then due to #42.

**Comment by r7vme** [comment]
Closed in favor https://github.com/giantswarm/versionbundle/pull/42

**Pull request closed without merging**